### PR TITLE
Feature/2019 04 14 Istio Service Mesh: service to service communication

### PR DESCRIPTION
**Pull request by pietervincken**

### CODE REVIEW

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+When a binding is created, the identities connected to it are allowed the access specified in the referenced service role.
+
+## Show me the code
+
+Explaining the Istio service to service authentication and authorization will now be explain by using an example setup.
+Note that the code snippets have been shortened in this blogpost.
+This is denoted with three dots `...`.
+The full examples can be found in the accompanying repository on [Github](https://github.com/pietervincken/istio-service-to-service-demo){:target="_blank" rel="noopener noreferrer"}
+
+### Prerequisits
```
Comment (Turbots) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]: Prerequisites

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+spec:
+  host: "*.local"
+  trafficPolicy:
+    tls:
+      mode: ISTIO_MUTUAL
+```
+
+### Enabling Role Based Access Control (RBAC) on the services
+
+Services can now communicate securly over mTLS.
```
Comment (Turbots) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]: securely

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+metadata:
+  name: default
+spec:
+  mode: 'ON_WITH_INCLUSION'
+  inclusion:
+    namespaces: ['with-istio']
+```
+
+After this RBAC config is applied, requests to the test-app services will start failing again. 
+The test-app currently doesn't have a role attached to it's service account that allows it to access the Couchdb database.
```
Comment (Turbots) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]: its* service account

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+To increase the security even further, RBAC can be added to the services.
+RBAC allows for roles to be defined that specify access to specific services in the cluster. 
+By attaching these roles to service accounts (which are connected to services) services can be premitted to access specific other services. 
+This limits the reach a single service has in the cluster and therefor adheres to the least privileges principle.
+
+The following manifest defines an cluster RBAC configuration.
+Only one such a configuration can exist in the entire service mesh and it needs to have the name `default`.
+The mode `ON_WITH_INCLUSION` specifies that all subjects that are listed in the inclusion section need to have RBAC enabled. 
+These subject can be namespace and/or specific services.
+Specifying the namespace with-istio in the inclusion section, enables RBAC for all services in that namespace.
```
Comment (yannickdeturck) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]:
```suggestion
Specifying the namespace `with-istio` in the inclusion section, enables RBAC for all services in that namespace.
```

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+metadata:
+  name: default
+spec:
+  mode: 'ON_WITH_INCLUSION'
+  inclusion:
+    namespaces: ['with-istio']
+```
+
+After this RBAC config is applied, requests to the test-app services will start failing again. 
+The test-app currently doesn't have a role attached to its service account that allows it to access the Couchdb database.
```
Comment (yannickdeturck) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]:
```suggestion
The test-app currently doesn't have a role attached to its service account that allows it to access the CouchDB database. 
```

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+  mode: 'ON_WITH_INCLUSION'
+  inclusion:
+    namespaces: ['with-istio']
+```
+
+After this RBAC config is applied, requests to the test-app services will start failing again. 
+The test-app currently doesn't have a role attached to its service account that allows it to access the Couchdb database. 
+Therefor all requests to the service will be rejected with an HTTP error code of 403.
+
+The following manifest creates a role that allows access to the Couchdb service for all GET requests on any given path.
```
Comment (yannickdeturck) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]:
```suggestion
The following manifest creates a role that allows access to the CouchDB service for all GET requests on any given path.
```

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+  name: couchdb-role
+  namespace: with-istio
+spec:
+  rules:
+  - services: ["couchdb.with-istio.svc.cluster.local"]
+    methods: ["GET"]
+    paths: ['*']
+```
+
+So next, we create link the new role to the service account of the test-app service.
```
Comment (yannickdeturck) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]:
```suggestion
So next, we link the new role to the service account of the test-app service.
```

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+This is done through a service role binding.
+There are two sections to this binding: the role and the subjects.
+The role is the one that was created using the previous manifest. 
+The subjects can be any identity known to Istio. 
+In the demo scenario, only the services accounts are known.
```
Comment (yannickdeturck) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]:
```suggestion
In the demo scenario, only the service accounts are known.
```

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+  namespace: with-istio
+spec:
+  subjects:
+  - user: "cluster.local/ns/with-istio/sa/test-app-test-app-chart"
+  roleRef:
+    kind: ServiceRole
+    name: "couchdb-role"
+```
+
+After applying the last manifest, requests should again be authorized and allowed to Couchdb.
```
Comment (yannickdeturck) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]:
```suggestion
After applying the last manifest, requests should again be authorized and allowed to CouchDB.
```

File: _posts/2019-04-14-istio-service-mesh-s2s.md [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+
+## Conclusion
+
+This demo showed how Istio can be used to enable strong authentication and authorization for services in a Kubernetes cluster. 
+It does this by enabling mTLS connections between the Envoy sidecar proxies.
+By creating roles, access to services can be granted or blocked. 
+A role can be connected to a service account to allow access. 
+
+In a nutshell, Istio allows cluster admins to enable secure communication and strong authentication and authorization mechanisms on their Kubernetes cluster without having to manage all kinds of certificates, usernames and passwords.
```
Comment (yannickdeturck) [inline _posts/2019-04-14-istio-service-mesh-s2s.md]:
```suggestion
In a nutshell, Istio allows cluster admins to enable secure communication, and strong authentication and authorization mechanisms on their Kubernetes cluster without having to manage all kinds of certificates, usernames and passwords. 
```

**Comment by yannickdeturck** [comment]
@tomverelst, @TYsewyn, do you guys want another look at the post before we publish it?

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+Because certificates can have a longer life span, this is easier to manage, compared to the JWT setup. 
+Setting up certificates, adopting the services to use them and rotating them over time, can be very challenging.
+Every services needs to expose a secure endpoint and serve the correct certificate.
+This is a non-trivial step and can, again, introduce very hard to debug issues.
+
+This is exactly where Istio pops up in the game.
+It makes using certificates for service to service encryption and authentication easy.
+Especially when using the automatic proxy injection, it's almost as easy as flipping a switch.
+
+There are three main components responsible for making this possible in Istio: Citadel, Pilot and Mixer.
```
Comment (yannickdeturck) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: With Galley added, does that make it have four main components now?

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 12]
```diff
@@ -0,0 +1,431 @@
...
+Currently this is described by specifying the full internal DNS name of the service, the methods and the paths that the role can access.
+
+### Service Role Bindings
+
+[Service Role Bindings](https://istio.io/docs/concepts/security/#servicerole){:target="_blank" rel="noopener noreferrer"} are used to connect identities (service accounts) or identity properties (namespaces) to actual roles. 
+When a binding is created, the identities connected to it are allowed the access specified in the referenced service role.
+
+## Show me the code
+
+Explaining the Istio service to service authentication and authorization will now be explain by using an example setup.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: "Explaining ... will now be *explained* using an example setup."
Can you rephrase this sentence? Sounds weird to me using "explain" twice.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+First, a specific service account is created for CouchDB. 
+This is needed as Istio will use the service accounts in Kubernetes as its identities.
+The service account is linked to the podspec in the stateful set definition. 
+This way, it can be used by the Istio proxy later on.
+
+Secondly, the probes have been adapted to work in Istio.
+Since Istio intercepts all traffic in the pod, it will also intercept requests from the Kube API to the service.
+Since the demo setup requires mTLS to be used, the probes would fail because the Kube API doesn't use mTLS.
+Instead of using changing the probes, Istio now has the option to rewrite the probes during the automatic proxy injection.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: Should it be "using" or "changing" the probes?

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+This way, it can be used by the Istio proxy later on.
+
+Secondly, the probes have been adapted to work in Istio.
+Since Istio intercepts all traffic in the pod, it will also intercept requests from the Kube API to the service.
+Since the demo setup requires mTLS to be used, the probes would fail because the Kube API doesn't use mTLS.
+Instead of using changing the probes, Istio now has the option to rewrite the probes during the automatic proxy injection. 
+More information on the probes can be found in the [Istio Docs](https://istio.io/help/ops/setup/app-health-check/#probe-rewrite){:target="_blank" rel="noopener noreferrer"}.
+
+Note that no Istio specific configuration is required in the service manifests.
+This is possible because the demo profile automatically enables the sidecar injection for all namespaces.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: This is possible because the demo profile automatically enables the sidecar *injector* **and we enabled the injection on the `with-istio` namespace using the `istio-injection=enabled` label.**

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+
+### Enabling Role Based Access Control (RBAC) on the services
+
+Services can now communicate securely over mTLS.
+To increase the security even further, RBAC can be added to the services.
+RBAC allows for roles to be defined that specify access to specific services in the cluster. 
+By attaching these roles to service accounts (which are connected to services) services can be premitted to access specific other services. 
+This limits the reach a single service has in the cluster and therefor adheres to the least privileges principle.
+
+The following manifest defines an cluster RBAC configuration.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: defines *a* cluster RBAC configuration.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+Only one such a configuration can exist in the entire service mesh and it needs to have the name `default`.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: Such configuration can only exist once in the entire service mesh and it needs to have the name `default`.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+The mode `ON_WITH_INCLUSION` specifies that all subjects that are listed in the inclusion section need to have RBAC enabled. 
+These subject can be namespace and/or specific services.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: These *subjects* can be *namespaces* and/or specific services.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+              - curl
+              - http://localhost:5984/_up
+          resources:
+            ...
+```
+
+### Install the service: test-app
+
+A small NodeJS application was created for this demo.
+It exposes a HTTP GET endpoint which connects to the CouchDB database.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: It exposes *an* HTTP GET endpoint which connects to the CouchDB database.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+RBAC allows for roles to be defined that specify access to specific services in the cluster. 
+By attaching these roles to service accounts (which are connected to services) services can be premitted to access specific other services. 
+This limits the reach a single service has in the cluster and therefor adheres to the least privileges principle.
+
+The following manifest defines an cluster RBAC configuration.
+Only one such a configuration can exist in the entire service mesh and it needs to have the name `default`.
+The mode `ON_WITH_INCLUSION` specifies that all subjects that are listed in the inclusion section need to have RBAC enabled. 
+These subject can be namespace and/or specific services.
+Specifying the namespace `with-istio` in the inclusion section, enables RBAC for all services in that namespace.
+By default the RBAC configuration will reject all requests which don't have the proper access defined with a HTTP 403 error code.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: with *an* HTTP 403 error code.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 20]
```diff
@@ -0,0 +1,431 @@
...
+kind: ClusterRbacConfig
+metadata:
+  name: default
+spec:
+  mode: 'ON_WITH_INCLUSION'
+  inclusion:
+    namespaces: ['with-istio']
+```
+
+After this RBAC config is applied, requests to the test-app services will start failing again.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: Shouldn't this be "test-app service" or "test-app instances"?

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+```
+
+So next, we link the new role to the service account of the test-app service.
+This is done through a service role binding.
+There are two sections to this binding: the role and the subjects.
+The role is the one that was created using the previous manifest. 
+The subjects can be any identity known to Istio. 
+In the demo scenario, only the service accounts are known.
+Istio defines a service account as a user identity. 
+As with the services names, the service account reference needs to be the full reference scoped towards the cluster.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: As with the *service* names

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 22]
```diff
@@ -0,0 +1,431 @@
...
+  namespace: with-istio
+spec:
+  subjects:
+  - user: "cluster.local/ns/with-istio/sa/test-app-test-app-chart"
+  roleRef:
+    kind: ServiceRole
+    name: "couchdb-role"
+```
+
+After applying the last manifest, requests should again be authorized and allowed to CouchDB.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: That's not true. Your request is allowed to be directed to your CouchDB instance, but that doesn't mean that your service is also authorized against the CouchDB instance. (in this case it is because there is no security configured on your CouchDB instance.)

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 23]
```diff
@@ -0,0 +1,431 @@
...
+
+## Conclusion
+
+This demo showed how Istio can be used to enable strong authentication and authorization for services in a Kubernetes cluster.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: This only showcases the authentication between pods via mTLS and the authorization of the communication between those two pods. It doesn't showcase the authorization between two services on application level.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 24]
```diff
@@ -0,0 +1,431 @@
...
+This demo showed how Istio can be used to enable strong authentication and authorization for services in a Kubernetes cluster. 
+It does this by enabling mTLS connections between the Envoy sidecar proxies.
+By creating roles, access to services can be granted or blocked. 
+A role can be connected to a service account to allow access. 
+
+In a nutshell, Istio allows cluster admins to enable secure communication, and strong authentication and authorization mechanisms on their Kubernetes cluster without having to manage all kinds of certificates, usernames and passwords. 
+The application developers don't need to adopt their application in order to communicate securly in the cluster, nor do they have to change their deployment configuration to enable the service mesh.
+
+This blogpost only highlighted a portion of the features of Istio. 
+Security is only a part of the featureset.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: Security is only a part of the *feature set*.

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 11 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: @TYsewyn I will include Envoy. I left it out because it isn't part of the Istio project. The mTLS stuff is discussed further down in the post :)

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 25]
```diff
@@ -0,0 +1,431 @@
...
+This is exactly where Istio pops up in the game.
+It makes using certificates for service to service encryption and authentication easy.
+Especially when using the automatic proxy injection, it's almost as easy as flipping a switch.
+
+There are three main components responsible for making this possible in Istio: Citadel, Pilot and Mixer. 
+
+Citadel is Istio's fortress of trust.
+It manages all certificates and acts as a Root CA in the Istio setup.
+
+Galley is the main information manager.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 26]
```diff
@@ -0,0 +1,431 @@
...
+
+### Policies
+
+[Policies](https://istio.io/docs/concepts/security/#authentication-policies){:target="_blank" rel="noopener noreferrer"} are at the heart of the mTLS setup in Istio.
+They define when mTLS should be used and how.
+Policies can be scoped in two levels: mesh wide (Mesh Policies) and namespace wide.
+
+### Destination Rules
+
+[Destination rules](https://istio.io/docs/concepts/traffic-management/#destination-rules){:target="_blank" rel="noopener noreferrer"} are a set of rules that are evaluated when a services is called.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 12 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 13 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 14 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 16 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 18 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 20 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Good point. I'll change it to test-app instance because the pod will actually reject it, not the Kubernetes service and there is only 1 instance in the demo :)

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 22 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: `After applying the last manifest, requests should again be authorized and allowed to connect to the CouchDB instance.`

Would this be better? I understand the confusion and you're right, the app is not authorized in CouchDB but it's authorized to connect to it in the service mesh.

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 15 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 17 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 19 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 24 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 21 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 23 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated, please review ;)

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 27]
```diff
@@ -0,0 +1,428 @@
...
+Both of these types of authentication flows are application level authentication flows. 
+They allow services to determine what resources an end-user or service can access.
+Istios service to service role based acccess control (RBAC) is not on an application level but on an communication level.
+It specifies which services can connect and communicate with each other. 
+In order to achieve this, Istio connects an identity to each service in the mesh and allows it to authenticate itself.
+The requested service can use this identity to determine if the service is allowed to connect or not.  
+Istio makes use of proxies to handle all traffic (into and out of services) and using mutual trusted certificates to secure the connection and provide an identity to these proxies.
+When using the automatic proxy injection, enabling Istio's service to service RBAC mechanism is almost as easy as flipping a switch.
+
+There are five main components responsible for making this possible in Istio: Citadel, Pilot, Galley, Mixer and Envoy.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: Maybe emphasize the components in the next paragraphs by making it **bold**?

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 28]
```diff
@@ -0,0 +1,428 @@
...
+    tls:
+      mode: ISTIO_MUTUAL
+```
+
+### Enabling Role Based Access Control (RBAC) on the services
+
+Services can now communicate securely over mTLS.
+To increase the security even further, RBAC can be added to the services.
+RBAC allows for roles to be defined that specify access to specific services in the cluster. 
+By attaching these roles to service accounts (which are connected to services) services can be premitted to access specific other services.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: Typo: *permitted*

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 29]
```diff
@@ -0,0 +1,428 @@
...
+  roleRef:
+    kind: ServiceRole
+    name: "couchdb-role"
+```
+
+After applying the last manifest, requests should again be authorized and allowed to connect to the CouchDB instance.
+
+## Conclusion
+
+This demo showed how Istio can be used to secure communication between service using mTLS.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: ... between *services* using mTLS.

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 30]
```diff
@@ -0,0 +1,428 @@
...
+Moreover it showed how the service mesh level authentication can be used to grant or deny access to services in the mesh.
+A role can be connected to a service account to allow access. 
+Important to note is that the service mesh only allowes or denies traffic.
+It doesn't influence the application level access.
+
+In a nutshell, Istio allows cluster admins to enable secure communication, and strong authentication and authorization mechanisms on their Kubernetes cluster without having to manage all kinds of certificates, usernames and passwords. 
+The application developers don't need to adopt their application in order to communicate securly in the cluster, nor do they have to change their deployment configuration to enable the service mesh.
```
Comment (TYsewyn) [inline _posts/2019-04-24-istio-service-mesh-s2s.md]: Typo: *securely*

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 31]
```diff
@@ -0,0 +1,428 @@
...
+* [What is Istio?](#what-is-istio)
+* [Istio concepts](#istio-concepts)
+* [Show me the code](#show-me-the-code)
+* [Conclusion](#conclusion)
+
+## What is Istio?
+
+Istio is a service mesh created by Google, Lyft and IBM. 
+It aims to simplify some security and management aspects of a microservices software architecture.
+More information on Istio and its feature can be found in its [docs](https://istio.io/docs/){:target="_blank" rel="noopener noreferrer"}.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 32]
```diff
@@ -0,0 +1,428 @@
...
+In this blogpost we will highlight one of the key security features of Istio: service to service authentication and authorization.
+For the sake of simplicity, this post will focus on an Istio setup in Kubernetes.
+
+In a microservices architecture, managing access to services can be a challenging operation. 
+For end-user facing services, JWTs are used to add authenication information a request.
+They are used by the service to determine what end-user is making the request.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 33]
```diff
@@ -0,0 +1,428 @@
...
+These tokens can be generated based on information that the end-user provides to an identity provider.
+In most cases this information is a username and password, with some additional 2FA if possible.
+This setup can be achieved by using OpenId Connect as a protocol with the authorization code grant flow and an identity provider like Keycloak.
+
+When services communicate with each other, they also need to provide an identity to each other.
+A common option to do this is using by using client credentials grant flow of OpenId Connect.
+In this flow, a service provides a client credentials to the identity provider to generate a token.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 34]
```diff
@@ -0,0 +1,428 @@
...
+This token will be used to communicate to a service.
+
+Both of these types of authentication flows are application level authentication flows.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 35]
```diff
@@ -0,0 +1,428 @@
...
+This setup can be achieved by using OpenId Connect as a protocol with the authorization code grant flow and an identity provider like Keycloak.
+
+When services communicate with each other, they also need to provide an identity to each other.
+A common option to do this is using by using client credentials grant flow of OpenId Connect.
+In this flow, a service provides a client credentials to the identity provider to generate a token.
+This token will be used to communicate to a service.
+
+Both of these types of authentication flows are application level authentication flows. 
+They allow services to determine what resources an end-user or service can access.
+Istios service to service role based acccess control (RBAC) is not on an application level but on an communication level.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 27 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 28 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 29 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md (hunk 30 above)
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 36]
```diff
@@ -0,0 +1,428 @@
...
+
+In a microservices architecture, managing access to services can be a challenging operation. 
+For end-user facing services, JWTs are used to add authenication information a request.
+They are used by the service to determine what end-user is making the request.
+These tokens can be generated based on information that the end-user provides to an identity provider.
+In most cases this information is a username and password, with some additional 2FA if possible.
+This setup can be achieved by using OpenId Connect as a protocol with the authorization code grant flow and an identity provider like Keycloak.
+
+When services communicate with each other, they also need to provide an identity to each other.
+A common option to do this is using by using client credentials grant flow of OpenId Connect.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 37]
```diff
@@ -0,0 +1,428 @@
...
+More information on Istio and its feature can be found in its [docs](https://istio.io/docs/){:target="_blank" rel="noopener noreferrer"}.
+In this blogpost we will highlight one of the key security features of Istio: service to service authentication and authorization.
+For the sake of simplicity, this post will focus on an Istio setup in Kubernetes.
+
+In a microservices architecture, managing access to services can be a challenging operation. 
+For end-user facing services, JWTs are used to add authenication information a request.
+They are used by the service to determine what end-user is making the request.
+These tokens can be generated based on information that the end-user provides to an identity provider.
+In most cases this information is a username and password, with some additional 2FA if possible.
+This setup can be achieved by using OpenId Connect as a protocol with the authorization code grant flow and an identity provider like Keycloak.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

File: _posts/2019-04-24-istio-service-mesh-s2s.md [hunk 38]
```diff
@@ -0,0 +1,428 @@
...
+## What is Istio?
+
+Istio is a service mesh created by Google, Lyft and IBM. 
+It aims to simplify some security and management aspects of a microservices software architecture.
+More information on Istio and its feature can be found in its [docs](https://istio.io/docs/){:target="_blank" rel="noopener noreferrer"}.
+In this blogpost we will highlight one of the key security features of Istio: service to service authentication and authorization.
+For the sake of simplicity, this post will focus on an Istio setup in Kubernetes.
+
+In a microservices architecture, managing access to services can be a challenging operation. 
+For end-user facing services, JWTs are used to add authenication information a request.
```
[the remark this answers is not in the document]
Comment (pietervincken) [reply _posts/2019-04-24-istio-service-mesh-s2s.md]: Updated

**Comment by TYsewyn** [comment]
LGTM! The only thing to do is to change the date so it will be the last one in the list of blogposts. :)
